rust/gui-client/src-tauri/src/client/network_changes/windows.rs

```diff
@@ -429,3 +403,337 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
     unsafe { Com::CoGetApartmentType(&mut apt_type, &mut apt_qualifier) }?;
     Ok((apt_type, apt_qualifier))
 }
+
+mod async_dns {
+    use anyhow::{Context, Result};
+    use std::{ffi::c_void, ops::Deref, path::Path};
+    use tokio::sync::mpsc;
+    use windows::Win32::{
+        Foundation::{CloseHandle, BOOLEAN, HANDLE, INVALID_HANDLE_VALUE},
+        System::Registry,
+        System::Threading::{
+            CreateEventA, RegisterWaitForSingleObject, UnregisterWaitEx, INFINITE,
+            WT_EXECUTEINWAITTHREAD,
+        },
+    };
+    use winreg::RegKey;
+
+    /// Opens and returns the IPv4 and IPv6 registry keys
+    ///
+    /// `HKEY_LOCAL_MACHINE/SYSTEM/CurrentControlSet/Services/Tcpip[6]/Parameters/Interfaces`
+    fn open_network_registry_keys() -> Result<(RegKey, RegKey)> {
+        let hklm = winreg::RegKey::predef(winreg::enums::HKEY_LOCAL_MACHINE);
+        let path_4 = Path::new("SYSTEM")
+            .join("CurrentControlSet")
+            .join("Services")
+            .join("Tcpip")
+            .join("Parameters")
+            .join("Interfaces");
+        let path_6 = Path::new("SYSTEM")
+            .join("CurrentControlSet")
+            .join("Services")
+            .join("Tcpip6")
+            .join("Parameters")
+            .join("Interfaces");
+        let flags = winreg::enums::KEY_NOTIFY;
+        Ok((
+            hklm.open_subkey_with_flags(path_4, flags)?,
+            hklm.open_subkey_with_flags(path_6, flags)?,
+        ))
+    }
+
+    pub(crate) fn run_debug() -> Result<()> {
+        tracing_subscriber::fmt::init();
+        let rt = tokio::runtime::Runtime::new()?;
+
+        let mut listener = CombinedListener::new()?;
+
+        rt.block_on(async move {
+            loop {
+                tokio::select! {
+                    _r = tokio::signal::ctrl_c() => break,
+                    r = listener.notified() => r?,
+                }
```

Comment:
Personally, I am not touching `tokio::select` with a 10ft pole because it has so many "interesting" features. If you only have two futures, it would suggest using `futures::future::select` instead which is trivial to understand when you look at the implementation.

If it is possible to remove the runtime above, you can even remove the `ctrl_c` listener here and just make this fn a loop of listening to be notified and compose it with `ctrl_c` at the outer layer of your app. I find listening to `ctrl_c` in `main` less surprising than several layers deep here.

YMMV :)